dev/com.ibm.ws.security.jwt/src/com/ibm/ws/security/jwt/internal/ConsumerUtil.java

```diff
@@ -56,801 +57,852 @@
+        }
+        JwtConsumerBuilder builder = initializeJwtConsumerBuilderWithoutValidation(config);
+        JwtConsumer firstPassJwtConsumer = builder.build();
+        return firstPassJwtConsumer.process(jwtString);
+    }
+
+    protected void validateJwtContext(JwtContext jwtContext, JwtConsumerConfig config) throws Exception {
+        Key key = getSigningKey(config, jwtContext);
+        JwtClaims jwtClaims = jwtContext.getJwtClaims();
+
+        if (tc.isDebugEnabled()) {
+            Tr.debug(tc, "Key from config: " + key);
+        }
+
+        validateClaims(jwtClaims, jwtContext, config);
+        validateSignatureAlgorithmWithKey(config, key);
+
+        JwtConsumerBuilder consumerBuilder = initializeJwtConsumerBuilderWithValidation(config, jwtClaims, key);
+        JwtConsumer jwtConsumer = consumerBuilder.build();
+        processJwtContextWithConsumer(jwtConsumer, jwtContext);
+    }
+
+    JwtConsumerBuilder initializeJwtConsumerBuilderWithoutValidation(JwtConsumerConfig config) {
+        JwtConsumerBuilder builder = new JwtConsumerBuilder();
+        builder.setSkipAllValidators();
+        builder.setDisableRequireSignature();
+        builder.setSkipSignatureVerification();
+        builder.setAllowedClockSkewInSeconds((int) ((config.getClockSkew()) / 1000));
+        return builder;
+    }
+
+    JwtConsumerBuilder initializeJwtConsumerBuilderWithValidation(JwtConsumerConfig config, JwtClaims jwtClaims,
+            Key key) throws MalformedClaimException {
+        JwtConsumerBuilder builder = new JwtConsumerBuilder();
+        builder.setExpectedIssuer(jwtClaims.getIssuer());
+        builder.setSkipDefaultAudienceValidation();
+        builder.setRequireExpirationTime();
+        builder.setVerificationKey(key);
+        builder.setRelaxVerificationKeyValidation();
+        builder.setAllowedClockSkewInSeconds((int) (config.getClockSkew() / 1000));
+        return builder;
+    }
+
+    void validateClaims(JwtClaims jwtClaims, JwtContext jwtContext, JwtConsumerConfig config)
+            throws MalformedClaimException, InvalidClaimException, InvalidTokenException {
+        String issuer = config.getIssuer();
+        if (issuer == null) {
+            issuer = mpConfigProps.get(MpConfigProperties.ISSUER);
+        }
+
+        long clockSkew = config.getClockSkew();
+        /**
+         * If clockSkew from server.xml is negative, then take the value of
+         * clock_skew from mpConfigProps
+         */
+        if (clockSkew < 0) {
+            String value = mpConfigProps.get(MpConfigProperties.CLOCK_SKEW);
+            if (value != null) {
+                clockSkew = Long.valueOf(value);
+            } else {
+                clockSkew = 0;
+            }
+        }
+
+        long tokenAge = 0;
+        // Take tokenAge value from server.xml
+        tokenAge = config.getTokenAge();
+        /**
+         * If tokenAge from server.xml is zero, then take the value of
+         * clock_skew from mpConfigProps
```

Comment:
"clock_skew" might have been a copy/paste error. We can change this to "tokenAge" some time later.